src/core/src/core_logic/PatchInstaller.py

```diff
@@ -283,7 +333,143 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
+        dependencies = package_manager.get_dependent_list(package_and_dependencies)
+
+        for package in dependencies:
+            if package not in all_packages:
+                continue
+            package_and_dependencies.append(package)
+            version = package_versions[packages.index(package)] if package in packages else Constants.DEFAULT_UNSPECIFIED_VALUE
+            package_and_dependency_versions.append(version)
+
+        for package in packages_in_batch:
+            package_manager.add_arch_dependencies(package_manager, package, packages, package_versions, package_and_dependencies, package_and_dependency_versions)
+
+        package_and_dependencies, package_and_dependency_versions = package_manager.dedupe_update_packages(package_and_dependencies, package_and_dependency_versions)
+
+        self.composite_logger.log("Packages including dependencies are: " + str(package_and_dependencies))
+
+    def install_packages_in_batches(self, all_packages, packages, package_versions, maintenance_window, package_manager, simulate=False):
+        number_of_batches = int(math.ceil(len(packages) / float(Constants.MAX_BATCH_SIZE_FOR_PACKAGES)))
+        self.composite_logger.log("\nDividing package install in batches. \nNumber of packages to be installed: " + str(len(packages)) + "\nBatch Size: " + str(Constants.MAX_BATCH_SIZE_FOR_PACKAGES) + "\nNumber of batches: " + str(number_of_batches))
+        installed_update_count = 0
+        patch_installation_successful = True
+        maintenance_window_batch_cutoff_reached = False
+
+        for batch_index in range(0, number_of_batches):
+            per_batch_installation_stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
+            per_batch_installation_stopwatch.start()
+
+            # Extension state check
+            if self.lifecycle_manager is not None:
+                self.lifecycle_manager.lifecycle_status_check()
+
+            begin_index = batch_index * Constants.MAX_BATCH_SIZE_FOR_PACKAGES
+            end_index = begin_index + Constants.MAX_BATCH_SIZE_FOR_PACKAGES - 1
+            end_index = min(end_index, len(packages) - 1)
+
+            packages_in_batch = []
+            package_versions_in_batch = []
+            skip_packages = []
+            already_installed_packages = []
+
+            for index in range(begin_index, end_index + 1):
+                if package_versions[index] == Constants.UA_ESM_REQUIRED:
+                    skip_packages.append(packages[index])
+                    self.status_handler.set_package_install_status(package_manager.get_product_name(packages[index]), str(package_versions[index]), Constants.NOT_SELECTED)
+                elif packages[index] not in self.last_still_needed_packages:
+                    # Should have got installed as dependent package of some other package. Package installation status should also have been set.
+                    already_installed_packages.append(package)
```

Comment:
package is used as local variable in this function even though it is not assigned any value yet in this function.
The statement should be already_installed_packages.append(packages[index])
I need to investigate why the test case related to this is succeeding. The test case is test_dependency_installed_successfully

Reply:
What was the finding on this? Also, I agree it should be already_installed_packages.append(packages[index])

Reply:
@GAURAVRAMRAKHYANI ^^ above question

Reply:
I have updated from package to packages[index]

@rane-rajasi  The reason why the test case test_dependency_installed_successfully was succeeding is due to a coincident. The variable package is being set below in this method. It was coincident that the value of the package was set to same as packages[index] here. Also, the variable package was not referred before setting below in this method so there were no errors.
The index is 3 and the value of packages[index] is grub-efi-amd64-bin